vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt

```diff
@@ -166,280 +159,262 @@ class OidcSiopWallet(
+            state = authenticationResponsePreparationState.request.parameters.state,
             vpToken = vpToken,
             presentationSubmission = presentationResultContainer?.presentationSubmission,
         )
         return KmmResult.success(
-            AuthenticationResponse(parameters, clientMetadata, jsonWebKeySet)
+            authenticationResponseParameters
         )
     }
 
     private fun Holder.PresentationResponseParameters.verifyFormatSupport(supportedFormats: FormatHolder) =
-        presentationSubmission.descriptorMap?.mapIndexed { _, descriptor ->
+        presentationSubmission.descriptorMap?.forEach { descriptor ->
             if (supportedFormats.isMissingFormatSupport(descriptor.format)) {
                 Napier.w("Incompatible JWT algorithms for claim format ${descriptor.format}: $supportedFormats")
                 throw OAuth2Exception(Errors.REGISTRATION_VALUE_NOT_SUPPORTED)
             }
         }
 
     private suspend fun buildPresentation(
-        params: AuthenticationRequestParametersFrom<*>,
-        audience: String,
-        presentationDefinition: PresentationDefinition,
-        clientMetadata: RelyingPartyMetadata
+        authenticationResponsePreparationState: AuthenticationResponsePreparationState,
+        presentationPreparationState: PresentationPreparationState,
+        inputDescriptorCredentialSubmissions: Map<String, InputDescriptorCredentialSubmission>?,
     ): Holder.PresentationResponseParameters {
-        if (params.parameters.nonce == null) {
-            Napier.w("nonce is null in ${params.parameters}")
-            throw OAuth2Exception(Errors.INVALID_REQUEST)
+        val credentialSubmissions = inputDescriptorCredentialSubmissions
+            ?: presentationPreparationState.inputDescriptorMatches.toDefaultSubmission()
+
+        presentationPreparationState.presentationSubmissionValidator.isValidSubmission(
+            credentialSubmissions.keys
+        ).also { isValidSubmission ->
+            if (!isValidSubmission) {
+                Napier.w("submission requirements are not satisfied")
+                throw OAuth2Exception(Errors.USER_CANCELLED)
+            }
         }
+
         return holder.createPresentation(
-            challenge = params.parameters.nonce,
-            audienceId = audience,
-            presentationDefinition = presentationDefinition,
-            fallbackFormatHolder = presentationDefinition.formats ?: clientMetadata.vpFormats,
-        ).getOrElse {
-            Napier.w("Could not create presentation", it)
+            challenge = authenticationResponsePreparationState.nonce,
+            audienceId = authenticationResponsePreparationState.audience,
+            presentationDefinitionId = presentationPreparationState.presentationDefinitionId,
+            presentationSubmissionSelection = credentialSubmissions,
+        ).getOrElse { exception ->
+            Napier.w("Could not create presentation: ${exception.message}")
```

Comment:
Why change it from passing the exception it self (as `it` previously)? Then it should print the stack trace and so on.

Reply:
Unsure why I changed that,
fixed in commit 990b482db2b4b3e6b5cb921e9cd557f2f12218e2